### PR TITLE
Multiplex Proxy SSH port

### PR DESCRIPTION
**Pull request by rosstimothy**
Adds support for multiplexing the Proxy SSH listener so that SSH and gRPC can be served over the listener at the same time.

Implements part of #19812.

### CODE REVIEW

**Review by ryanclark — APPROVED** [review]

File: lib/service/service.go [hunk 1]
```diff
@@ -3713,7 +3713,33 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
...
+				utils.GRPCServerStreamErrorInterceptor,
+				otelgrpc.StreamServerInterceptor(),
+			),
+			grpc.Creds(credentials.NewTLS(serverTLSConfig)),
+		)
+
+		// start grpc server
+		go func() {
+			if err := grpcServer.Serve(listeners.sshGrpc); !errors.Is(err, grpc.ErrServerStopped) {
+				log.WithError(err).Warn("SSH gRPC server terminated unexpectedly")
```
Comment (codingllama) [inline lib/service/service.go:3739]: Possibly?

```suggestion
				log.WithError(err).Error("SSH gRPC server terminated unexpectedly")
```

File: lib/service/service.go [hunk 2]
```diff
@@ -2952,6 +2954,7 @@ type proxyListeners struct {
 	mux              *multiplexer.Mux
 	tls              *multiplexer.WebListener
 	ssh              net.Listener
+	sshGrpc          net.Listener
```
Comment (codingllama) [inline lib/service/service.go:2957]: nit:

```suggestion
	sshGRPC          net.Listener
```

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: lib/service/service.go [hunk 3]
```diff
@@ -4323,6 +4349,16 @@ func setupALPNRouter(listeners *proxyListeners, serverTLSConfig *tls.Config, cfg
 	})
 	listeners.ssh = sshProxyListener
 
+	sshGrpcListener := alpnproxy.NewMuxListenerWrapper(listeners.sshGrpc, listeners.web)
```
Comment (codingllama) [inline lib/service/service.go:4352]: nit:

```suggestion
	sshGRPCListener := alpnproxy.NewMuxListenerWrapper(listeners.sshGrpc, listeners.web)
```

**Review by codingllama — COMMENTED** [review]
Apologies for the delay.

Do we have a test for connections over the new channel? Would it belong here?

File: lib/service/service.go (hunk 2 above)
Comment (zmb3) [reply lib/service/service.go:2957]: I might also add a bit more info here. Something like:

```
// sshGRPC listens for traffic intended for the proxy.
// The SSH traffic is encapsulated in a GRPC stream to
// avoid the cost of an additional SSH handshake.
```

File: integration/integration_test.go [hunk 4]
```diff
@@ -6700,3 +6704,100 @@ func testSFTP(t *testing.T, suite *integrationTestSuite) {
...
+			// create an authenticated client for the user
+			tc, err := rc.NewClient(helpers.ClientConfig{
+				Login:   me.Username,
+				Cluster: helpers.Site,
+				Host:    Host,
+			})
+			require.NoError(t, err)
+
+			// connect via SSH
+			ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
```
Comment (codingllama) [inline integration/integration_test.go:6779]:
```suggestion
			ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
			defer cancel()
```

so it doesn't hang around on errors.

File: integration/integration_test.go [hunk 5]
```diff
@@ -6700,3 +6704,100 @@ func testSFTP(t *testing.T, suite *integrationTestSuite) {
...
+			ctx, cancel = context.WithTimeout(context.Background(), 15*time.Second)
+			defer cancel()
+			conn, err := grpc.DialContext(
+				ctx,
+				tc.SSHProxyAddr,
+				grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
+				grpc.WithBlock(),
+			)
+			require.NoError(t, err)
+			require.Equal(t, connectivity.Ready, conn.GetState())
```
Comment (codingllama) [inline integration/integration_test.go:6799]: Won't this be timing sensitive or flaky? Could we exercise the connection somehow instead?

Exercising the connection above would be good too.

File: integration/integration_test.go [hunk 6]
```diff
@@ -6700,3 +6704,100 @@ func testSFTP(t *testing.T, suite *integrationTestSuite) {
...
+	tests := []struct {
+		name              string
+		disableTLSRouting bool
+	}{
+		{
+			name: "TLS routing enabled",
+		},
+		{
+			name:              "TLS routing disabled",
+			disableTLSRouting: true,
```
Comment (codingllama) [inline integration/integration_test.go:6722]: Do you have plans to add a test without disableTLSRouting? Should we leave a TODO (or drop the param)?

**Comment by codingllama** [comment]
Thanks for adding a test!

File: integration/integration_test.go (hunk 6 above)
Comment (codingllama) [reply integration/integration_test.go:6722]: 🤦

File: integration/integration_test.go (hunk 5 above)
Comment (codingllama) [reply integration/integration_test.go:6799]: Fair enough.

**Review by codingllama — APPROVED** [review]

**Pull request merged**